core/controller/src/main/scala/whisk/core/entitlement/KindRestrictor.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.entitlement
+
+import whisk.common.Logging
+import whisk.common.TransactionId
+import whisk.core.ConfigKeys
+import whisk.core.entity.{Exec, Identity}
+import pureconfig.loadConfigOrThrow
+
+/**
+ * Restrict allowed kinds via provided whitelists
+ *
+ * @param whitelist set of default allowed kinds when not explicitly available via namespace limits
+ */
+protected[core] case class DefaultAllowedKinds(whitelist: Option[Set[String]] = None)
+
+/**
+ * A class for determining whether a given namespace is allowed to contain a given kind.
+ *
+ * @param allowedKinds a default set of allowed kinds
+ */
+class KindRestrictor(allowedKinds: DefaultAllowedKinds = loadConfigOrThrow[DefaultAllowedKinds](ConfigKeys.runtimes))(
+  implicit logging: Logging) {
+
+  logging.info(this, s"allowedKinds = $allowedKinds")(TransactionId.controller)
+
+  private def execKindAllowed(exec: Option[Exec], kinds: Set[String]): Boolean = {
+    exec.map(e => kinds.contains(e.kind)).getOrElse(false)
+  }
+
+  /**
+   * Checks whether a given kind is allowed for a given user.
+   */
+  def check(user: Identity, exec: Option[Exec])(implicit tid: TransactionId): Boolean = {
```

Comment:
I believe the intent here to entitle a subject to a runtime if:
  1. they are explicitly granted rights to it in their identity record, or
  2. the runtime is whitelisted

The implementation assumes if there are no white listed runtimes (ie == `None`), then all runtimes are white listed.
Is that a correct understanding of the intent?

Reply:
Yes, that is correct. And similarly, if an empty set is provided as a whitelist, it means access to no runtimes unless explicitly set for a subject